src/gui/tagfiltermodel.cpp

```diff
@@ -0,0 +1,326 @@
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#include "tagfiltermodel.h"
+
+#include <QHash>
+#include <QIcon>
+
+#include "base/bittorrent/session.h"
+#include "base/bittorrent/torrenthandle.h"
+#include "guiiconprovider.h"
+
+class TagModelItem
+{
+public:
+    TagModelItem(const QString &tag, const QString &displayName, int torrentsCount = 0)
+        : m_tag(tag)
+        , m_displayName(displayName)
+        , m_torrentsCount(torrentsCount)
+    {
+    }
+
+    ~TagModelItem() {}
+
+    QString tag() const
+    {
+        return m_tag;
+    }
+
+    // This is the string that actually shows up in the widget. Usually it is the same
+    // as the tag, except for special tags like All and Untagged.
+    QString displayName() const
+    {
+        return m_displayName;
+    }
+
+    int torrentsCount() const
+    {
+        return m_torrentsCount;
+    }
+
+    void increaseTorrentsCount()
+    {
+        ++m_torrentsCount;
+    }
+
+    void decreaseTorrentsCount()
+    {
+        Q_ASSERT(m_torrentsCount > 0);
+        --m_torrentsCount;
+    }
+
+private:
+    QString m_tag;
+    QString m_displayName;
+    int m_torrentsCount;
+};
+
+TagFilterModel::TagFilterModel(QObject *parent)
+    : QAbstractItemModel(parent)
+{
+    using Session = BitTorrent::Session;
+    auto session = Session::instance();
+
+    connect(session, &Session::tagAdded, this, &TagFilterModel::tagAdded);
+    connect(session, &Session::tagRemoved, this, &TagFilterModel::tagRemoved);
+    connect(session, &Session::torrentTagAdded, this, &TagFilterModel::torrentTagAdded);
+    connect(session, &Session::torrentTagRemoved, this, &TagFilterModel::torrentTagRemoved);
+    connect(session, &Session::torrentAdded, this, &TagFilterModel::torrentAdded);
+    connect(session, &Session::torrentAboutToBeRemoved, this, &TagFilterModel::
+torrentAboutToBeRemoved);
+
+    populate();
+}
+
+TagFilterModel::~TagFilterModel()
+{
+    foreach (const TagModelItem *item, m_tagItems.values())
+        delete item;
+}
+
+bool TagFilterModel::isSpecialItem(const QModelIndex &index)
+{
+    // the first two items are special items: 'All' and 'Untagged'
+    return (!index.parent().isValid() && (index.row() <= 1));
+}
+
+int TagFilterModel::columnCount(const QModelIndex &) const
+{
+    return 1;
+}
+
+QVariant TagFilterModel::data(const QModelIndex &index, int role) const
+{
+    if (!index.isValid() || index.column() != 0)
+        return QVariant();
+
+    auto item = static_cast<TagModelItem *>(index.internalPointer());
+
+    if (role == Qt::DecorationRole)
+        return GuiIconProvider::instance()->getIcon("inode-directory");
```

Comment:
Once I did a clean build, I noticed that this change caused my icons to disappear.

![tags-ui-update](https://user-images.githubusercontent.com/6699718/27360702-5aa01fa0-55e8-11e7-831b-3bfedb7d1cd9.png)

This is with a stock Linux Mint 18.1. I am not sure why it is not using the 'fallback' icon if the tags icons are not there. But I decided to revert back to the original icons I was using to avoid any problems.